Topics/Tech_Stacks/cuda.md

```diff
@@ -0,0 +1,260 @@
+
+## Installation:
+
+Note that since CUDA is a proprietary technology, you MUST have an NVIDIA GPU with a valid CUDA compute capability (see [here](https://developer.nvidia.com/cuda-gpus)). You can also install this on a machine with an NVIDIA GPU installed, such as a UofT lab machine, or a computer you can SSH into with an NVIDIA GPU.
+
+To install CUDA on your system, simply download the [CUDA Toolkit](https://developer.nvidia.com/cuda-downloads) onto your computer. Linux users can either install the executable directly from the NVIDIA website or they can use their favourite package manager to get it from the AUR, APT, and so on. Windows users can simply download the executable directly from the CUDA Toolkit website linked below:
+
+https://developer.nvidia.com/cuda-downloads
+
+Information retrieved from: 
+
+https://docs.nvidia.com/cuda/cuda-installation-guide-linux/index.html and 
+https://docs.nvidia.com/cuda/cuda-installation-guide-microsoft-windows/index.html
+
+## A brief (oversimplified) primer to GPU architecture:
+
+Let's assume our GPU is an NVIDIA GTX 1060. A GTX 1060 has 10 streaming multiprocessors, 15 blocks with a total of 1024 threads per block. What does this all mean? Each streaming multiprocessor (SM) contains a certain amount of blocks it can run. A block is an organization of threads, in our case a block contains up to 1024 threads. Threads are run concurrently across blocks, which get scheduled onto SMs. Each SM can handle a certain number of blocks concurrently. In our case, with 10 SMs, if we have 10 blocks, one per SM, each block would be computed in parallel by a separate SM. However, if the number of blocks exceeds 10, let's say 15 blocks, the SMs would need to context switch between multiple blocks, potentially reducing performance due to switching overhead. Keep in mind that all these performance specifications vary between GPU. Some GPUs support multiple blocks running on an SM without context switching.
```

Comment:
I would consider an brief explanation of "context switch" beneficial to users who do not know about GPU architecture.